[clang][dataflow] Fully support Environment construction for Stmt analysis.

bazuzi opened the pull request:
Assume in fewer places that the analysis is of a FunctionDecl, and initialize the Environment properly for Stmts.

Moves constructors for Environment to header to make it more obvious that there are only minor differences between them and very little initialization in the constructors.

Tested with check-clang-tooling.

bazuzi committed:
[clang][dataflow] Fully support Environment construction for Stmt analysis.

Assume in fewer places that the analysis is of a FunctionDecl, and initialize the Environment properly for Stmts.

Moves constructors for Environment to header to make it more obvious
that there are only minor differences between them and very little
initialization in the constructors.

Tested check-clang-tooling.

bazuzi commented:
@ymand Can you review and eventually merge for me?

bazuzi committed:
Add test and crash fix for analysis of Stmt.

ymand reviewed (commented):
I think we can do this without resorting to a pointer union and templating of the initialization functions, but it will take a little refactoring.

First, notice that `stmt` doesn't make sense for the CallStack. The call stack serves two purposes -- to avoid recursion (line 627) and to process the return val (lines 801-811).  Neither of these cases apply for stmts, as you've correctly identified in your modifications to the code (e.g. line 776 in your new version of DataflowEnvironment.cpp)

So, I think its fair to say that CallStack should remain as is. If you're analyzing a Stmt, then you simply don't push onto the call stack. (I'm actually thinking we should not push the starting FunctionDecl onto the call stack either, since it's not called from anywhere, but that's a separate issue...)

That brings us to the second use of the function decl -- initialization. here's where refactoring comes in -- if you look at the init call stack, it's actually doing two things: use the "function decl" stuff (like parameters) and using the body (which is just a stmt). Currently, these are nested. But, I think that, instead, you can refactor to sequence them. Then, for FunctionDecl cases we call both but for Stmts just the second. The way to distinguish I think is just to explicitly store both the FD and the Stmt as separate fields. Then, init will assume there's always a stmt but will check the FD for null. Or, it can rely on the call stack and just look at the top of the call stack, if any.

WDYT?

bazuzi committed:
Store the initial analysis target outside of the CallStack.

Reverts TargetStack to CallStack and stores the initial target statement
and, when present, its containing FunctionDecl separately from
CallStack.

bazuzi commented:
I agree with the usefulness of storing the initial target separately, reducing checks and branching based on whether the current target is a `Stmt` or `Function`. Since I was touching everywhere `CallStack` was used anyway, I went ahead with not pushing the starting `FunctionDecl` onto that stack.

I wasn't immediately able to reduce the `initialize` duplication for statements and function bodies since `buildResultObjectMap` and `initFieldsGlobalsAndFuncs`, via `getReferencedDecls`, both handle `FunctionDecl`s and `Stmt`s differently, pulling info from a `FunctionDecl` beyond just its body. If templating `initFieldsGlobalsAndFuncs` is really undesirable, I could either pass in the `ReferencedDecls` from current callers or insert an intermediate with two overloads that takes the `Stmt`/`FunctionDecl` and passes along the `ReferencedDecls`.

ymand commented:
> I agree with the usefulness of storing the initial target separately, reducing checks and branching based on whether the current target is a `Stmt` or `Function`. Since I was touching everywhere `CallStack` was used anyway, I went ahead with not pushing the starting `FunctionDecl` onto that stack.

Thanks!

> I wasn't immediately able to reduce the `initialize` duplication for statements and function bodies since `buildResultObjectMap` and `initFieldsGlobalsAndFuncs`, via `getReferencedDecls`, both handle `FunctionDecl`s and `Stmt`s differently, pulling info from a `FunctionDecl` beyond just its body. If templating `initFieldsGlobalsAndFuncs` is really undesirable, I could either pass in the `ReferencedDecls` from current callers or insert an intermediate with two overloads that takes the `Stmt`/`FunctionDecl` and passes along the `ReferencedDecls`.

I'm generally allergic to templates unless necessary, but that doesn't mean I'm right :) I'm inclined towards either of the solutions you suggested involving passing `ReferencedDecls` (with a slight, arbitrary preference for the first since it involves fewer new function declarations), but I leave it up to you since you're actually writing the code and have a better view as to which solution is cleaner. Also happy to hear from other reviewers.

bazuzi committed:
Account for changes to stack size and initialize when analyzing Stmts.

bazuzi committed:
Un-template initFieldsGlobalsAndFuncs and update ResultObjectVisitor arg
and comments.

ymand commented on clang/lib/Analysis/FlowSensitive/TypeErasedDataflowAnalysis.cpp:
```diff
@@ -476,7 +476,7 @@ runTypeErasedDataflowAnalysis(
 
   std::optional<Environment> MaybeStartingEnv;
-  if (InitEnv.callStackSize() == 1) {
+  if (InitEnv.callStackSize() == 0) {
```
aside: this line now feels more "right" than the original version.

ymand reviewed (approved):
Nice!

bazuzi committed:
Put back the correct name of CallStack in comment.

martinboehme commented on clang/unittests/Analysis/FlowSensitive/DataflowEnvironmentTest.cpp:
```diff
@@ -403,4 +405,35 @@ TEST_F(EnvironmentTest,
+  auto *DeclStatement = const_cast<DeclStmt *>(
+      selectFirst<DeclStmt>("d", match(declStmt().bind("d"), Context)));
+  ASSERT_THAT(DeclStatement, NotNull());
+  auto *ConstructExpr = selectFirst<CXXConstructExpr>(
+      "c", match(cxxConstructExpr().bind("c"), Context));
+  ASSERT_THAT(ConstructExpr, NotNull());
```
Instead of matching for a `CXXConstructExpr`, just grab the `DeclStatement->getInit()`?

This would seem to be simpler, clearer, and less brittle (e.g. if you compile this as C++17, there isn't a `CXXConstructExpr` in the AST).

martinboehme commented on clang/unittests/Analysis/FlowSensitive/DataflowEnvironmentTest.cpp:
```diff
@@ -403,4 +405,35 @@ TEST_F(EnvironmentTest,
+  // Don't crash when initializing.
+  Env.initialize();
+  // And don't crash when retrieving the result object location.
+  Env.getResultObjectLocation(*ConstructExpr);
```
Would be nicer to have this test do more than simply test that we don't crash -- but I'm not sure if this is possible. (IIUC, we don't model the storage location for `AnS`, correct?)

bazuzi replied:
Correct, only result objects and referenced decls. I didn't see much opportunity to check the value of anything in particular.

martinboehme commented on clang/unittests/Analysis/FlowSensitive/TypeErasedDataflowAnalysisTest.cpp:
```diff
@@ -146,6 +146,30 @@ TEST_F(DataflowAnalysisTest, DiagnoseFunctionDiagnoserCalledOnEachElement) {
 
+TEST_F(DataflowAnalysisTest, CanAnalyzeStmt) {
+  std::string Code = R"cc(
+      struct S {int i;};
+      S getAnS() {return S{1};};
+      void foo() {
+        S AnS = getAnS();
+      }
```
```suggestion
      struct S { int i; };
      S getAnS() {return S{1};};
      S AnS = getAnS();
```

bazuzi replied:
Same reason not to use a global variable as above. Additionally, `AdornedCFG::build` expects to take a `FunctionDecl` in which the provided `Stmt` resides. In the case of `DeclStmt`s, this `FunctionDecl` doesn't seem to actually be used at all, but support may not be guaranteed.

martinboehme replied:
I see, thanks for the explanation.

martinboehme commented on clang/unittests/Analysis/FlowSensitive/TypeErasedDataflowAnalysisTest.cpp:
```diff
@@ -146,6 +146,30 @@ TEST_F(DataflowAnalysisTest, DiagnoseFunctionDiagnoserCalledOnEachElement) {
+      }
+    )cc";
+  AST = tooling::buildASTFromCodeWithArgs(Code, {"-std=c++11"});
+  const auto &DeclStatement = matchNode<DeclStmt>(declStmt());
```
As above: Match the name of the declaration?

martinboehme commented on clang/unittests/Analysis/FlowSensitive/TypeErasedDataflowAnalysisTest.cpp:
```diff
@@ -146,6 +146,30 @@ TEST_F(DataflowAnalysisTest, DiagnoseFunctionDiagnoserCalledOnEachElement) {
+TEST_F(DataflowAnalysisTest, CanAnalyzeStmt) {
+  std::string Code = R"cc(
+      struct S {int i;};
+      S getAnS() {return S{1};};
```
Why do we need this function instead of simply initializing `AnS` directly with `S{1}` as in the test above?

bazuzi replied:
Just an ever so slightly more interesting analysis to test. Simplified now so that we can check the field value.

martinboehme commented on clang/unittests/Analysis/FlowSensitive/TypeErasedDataflowAnalysisTest.cpp:
```diff
@@ -146,6 +146,30 @@ TEST_F(DataflowAnalysisTest, DiagnoseFunctionDiagnoserCalledOnEachElement) {
+  Environment Env(*DACtx, const_cast<DeclStmt &>(DeclStatement));
+
+  ASSERT_THAT_ERROR(runDataflowAnalysis(*ACFG, Analysis, Env).takeError(),
+                    llvm::Succeeded());
```
Again, can we make this test stronger than simply asserting that we don't error out?

For example, if we made the member variable of `S` a bool instead of an integer, we could verify that it was initialized with a specific boolean literal.

bazuzi replied:
Changed to bool and added a check.

bazuzi committed:
Respond to review comments.

Mostly improvements to tests and comments, plus an assertion and
explicit type.

bazuzi commented on clang/unittests/Analysis/FlowSensitive/DataflowEnvironmentTest.cpp:
```diff
@@ -403,4 +405,35 @@ TEST_F(EnvironmentTest,
+
+  std::string Code = R"cc(
+      struct S {int i;};
+      void foo() {
+        S AnS = S{1};
+      }
```
A variable declaration outside a function doesn't actually exist within a Stmt, nor is it a Stmt.

In order to analyze a global variable declaration, one option is to synthesize a DeclStmt that holds the VarDecl. I was trying to keep this test simple, but could add an additional test for this synthesis pattern?

martinboehme replied:
> A variable declaration outside a function doesn't actually exist within a Stmt, nor is it a Stmt.

But isn't this the use case for which you're making the change in this PR?

I don't see why you need the variable declaration to be a `Stmt` though. In any case, you're only analyzing the initializer, not the `DeclStmt` itself. Instead of searching for the `DeclStmt`, you could search for the `VarDecl` and then initialize the `Environment` with the `getInit()` of that `VarDecl`. This wouldn't make the test any more complicated and would be equivalent to what you're testing here?

martinboehme replied:
Having looked at this again, I now realize that you're analyzing the `DeclStmt` itself, and not just the declaration's initializer. I think analyzing just the initializer would have the same effect for the purposes of this test -- but either of these two options seems fine, so feel free to leave this as is.

bazuzi commented on clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp:
```diff
@@ -290,13 +295,13 @@ widenKeyToValueMap(const llvm::MapVector<Key, Value *> &CurMap,
 public:
   // `ResultObjectMap` will be filled with a map from record prvalues to result
-  // object. If the function being analyzed returns a record by value,
+  // object. If the sub-tree to be visited returns a record by value,
```
Same confusion around analysis as above. I'm also having difficulty with the wording for this because this is a comment on the constructor, not a Visit/Traverse function. Tried another version.

bazuzi commented on clang/lib/Analysis/FlowSensitive/DataflowEnvironment.cpp:
```diff
@@ -290,13 +295,13 @@ widenKeyToValueMap(const llvm::MapVector<Key, Value *> &CurMap,
-// This traverses the body of the function to be analyzed; for each result
-// object that it encounters, it propagates the storage location of the result
-// object to all record prvalues that can initialize it.
+// This traverses the AST sub-tree to be visited; for each result object that
```
I found "analyzed" to be confusing because it seemed to imply that this Visitor was doing the analyzing, which within Environment typically refers to a dataflow analysis, not any generic analysis of something.

Looks like all starting calls to use this Visitor use a Traverse* method, so the fact that it traverses seems obvious enough that we can just remove this clause.

martinboehme replied:
> I found "analyzed" to be confusing because it seemed to imply that this Visitor was doing the analyzing, which within Environment typically refers to a dataflow analysis, not any generic analysis of something.

Ah, understood. What I meant when I wrote the original comment was indeed that the visitor traverses the body of the function on which we later want to perform dataflow analysis. But I see how this can be confusing.

Agree that this bit can simply be dropped; "this traverses the AST sub-tree to be visited" is true of any `RecursiveASTVisitor`.

bazuzi commented on clang/include/clang/Analysis/FlowSensitive/DataflowEnvironment.h:
```diff
@@ -164,23 +169,28 @@ class Environment {
+  ///
+  /// If the target is a non-static member function, initializes the environment
+  /// with a symbolic representation of the `this` pointee.
   void initialize();
```
The call stack being assumed non-empty was the case previously. Now these two are not technically on the call stack, and the function will no longer crash if called without InitialTargetStmt being non-null. It just doesn't do anything. Updated the comment to indicate this.

bazuzi commented on clang/include/clang/Analysis/FlowSensitive/DataflowEnvironment.h:
```diff
@@ -718,16 +730,24 @@ class Environment {
                         ArrayRef<const Expr *> Args);
 
+  // FIXME: Add support for resetting globals after function calls to enable
+  // the implementation of sound analyses.
```
Thanks for catching, there was some iteration on the placement of the implementation of `initFieldsGlobalsAndFuncs` and I forgot to move this back.
